src/platforms/atomic-kms/server/kms/quirks.cpp

```diff
@@ -0,0 +1,204 @@
+            }();
+        mir::log_debug("Quirks: checking device with devnode: %s, driver %s", device.devnode(), driver);
+        bool const should_skip_driver = drivers_to_skip.count(driver);
+        bool const should_skip_devnode = devnodes_to_skip.count(devnode);
+        if (should_skip_driver)
+        {
+            mir::log_info("Quirks: skipping device %s (matches driver quirk %s)", devnode, driver);
+        }
+        if (should_skip_devnode)
+        {
+            mir::log_info("Quirks: skipping device %s (matches devnode quirk %s)", devnode, devnode);
+        }
+        return should_skip_driver || should_skip_devnode;
+    }
+
+    auto require_modesetting_support(mir::udev::Device const& device) const -> bool
+    {
+        auto const devnode = value_or(device.devnode(), "");
+        auto const parent_device = device.parent();
+        auto const driver =
+            [&]()
+            {
+                if (parent_device)
+                {
+                    return value_or(parent_device->driver(), "");
+                }
+                mir::log_warning("udev device has no parent! Unable to determine driver for quirks.");
+                return "<UNKNOWN>";
+            }();
+        mir::log_debug("Quirks: checking device with devnode: %s, driver %s", device.devnode(), driver);
+
+        bool const should_skip_modesetting_support = skip_modesetting_support.count(driver);
+        if (should_skip_modesetting_support)
+        {
+            mir::log_info("Quirks: skipping modesetting check %s (matches driver quirk %s)", devnode, driver);
+        }
+        return !should_skip_modesetting_support;
+    }
+
+private:
+    /* AST is a simple 2D output device, built into some motherboards.
+     * They do not have any 3D engine associated, so were quirked off to avoid https://github.com/canonical/mir/issues/2678
+     *
+     * At least as of drivers ≤ version 550, the NVIDIA atomic implementation is buggy in a way that prevents
+     * Mir from working. Quirk off atomic-kms on NVIDIA.
+     */
+    std::unordered_set<std::string> drivers_to_skip = { "nvidia", "ast" };
+    std::unordered_set<std::string> devnodes_to_skip;
+    // We know this is currently useful for virtio_gpu, vc4-drm and v3d
+    std::unordered_set<std::string> skip_modesetting_support = { "virtio_gpu", "vc4-drm", "v3d" };
+};
+
+mga::Quirks::Quirks(const options::Option& options)
+    : impl{std::make_unique<Impl>(options)}
+{
+}
+
+mga::Quirks::~Quirks() = default;
+
+auto mga::Quirks::should_skip(udev::Device const& device) const -> bool
+{
+    return impl->should_skip(device);
+}
+
+void mga::Quirks::add_quirks_option(boost::program_options::options_description& config)
+{
+    config.add_options()
+        (quirks_option_name,
+         boost::program_options::value<std::vector<std::string>>(),
+         "[platform-specific] Driver quirks to apply (may be specified multiple times; multiple quirks are combined)");
```

Comment:
This same option is added by `gbm-kms` - and that spells trouble if both platforms are available. (Similar to #885)

Reply:
Is this resolved by #3590?